src/shims/panic.rs

```diff
@@ -0,0 +1,180 @@
+use rustc::mir;
+use crate::*;
+use super::machine::FrameData;
+use rustc_target::spec::PanicStrategy;
+use crate::rustc_target::abi::LayoutOf;
+
+/// Holds all of the relevant data for a call to
+/// __rust_maybe_catch_panic
+///
+/// If a panic occurs, we update this data with
+/// the information from the panic site
+#[derive(Debug)]
+pub struct CatchUnwindData<'tcx> {
+    /// The 'data' argument passed to `__rust_maybe_catch_panic`
+    pub data: Pointer<Tag>,
+    /// The `data_ptr` argument passed to `__rust_maybe_catch_panic`
+    pub data_ptr: MPlaceTy<'tcx, Tag>,
+    /// The `vtable_ptr` argument passed to `__rust_maybe_catch_panic`
+    pub vtable_ptr: MPlaceTy<'tcx, Tag>,
+    /// The `dest` from the original call to `__rust_maybe_catch_panic`
+    pub dest: PlaceTy<'tcx, Tag>,
+    /// The `ret` from the original call to `__rust_maybe_catch_panic`
+    pub ret: mir::BasicBlock,
+}
+
+impl<'mir, 'tcx> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
+
+    /// Handles the special "miri_start_panic" intrinsic, which is called
+    /// by libpanic_unwind to delegate the actual unwinding process to Miri
+    #[inline(always)]
+    fn handle_miri_start_panic(
+        &mut self,
+        args: &[OpTy<'tcx, Tag>],
+        unwind: Option<mir::BasicBlock>
+    ) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+
+        trace!("miri_start_panic: {:?}", this.frame().span);
+
+        if this.tcx.tcx.sess.panic_strategy() == PanicStrategy::Abort  {
+            // FIXME: Actually print out the payload here
```

Comment:
Why should Miri be the thing that does the printing? Shouldn't we run the code that also does it at run-time when `panic=abort`?

Reply:
You're right - the panic hook takes care of this.